docs/sc-memory/api/cpp/guides/simple_guide_for_implementing_agent.md

```diff
@@ -0,0 +1,524 @@
+
+In addition to agents that initiate actions themselves and then perform these actions, there is a need to implement agents that perform actions initiated by other agents. For this class of agents, it is much easier to create a initial initiation construction in the knowledge base.
+
+--- 
+
+### **2. Create folder with source and header files for agent and `CMakeLists.txt.` file.**
+
+You should get something like this structure:
+
+```plaintext
+set-agents-module/
+├── CMakeLists.txt
+├── agents/
+│   ├── sc_agent_calculate_set_power.hpp
+│   └── sc_agent_calculate_set_power.сpp
+```
+
+---
+
+### **3. Write `CMakeLists.txt` file.**
+  
+The `CMakeLists.txt` file should describe a process of building your agent code. It should contain various instructions and parameters needed to compile and link agent source code to sc-machine libraries.
+
+**CMakeLists.txt**
+
+```cmake
+file(GLOB SOURCES CONFIGURE_DEPENDS
+    "*.cpp" "*.hpp"
+    "agents/*.cpp" "agents/*.hpp"
+)
+
+# Create and link your library with using libraries.
+add_library(set-agents SHARED ${SOURCES})
+target_link_libraries(set-agents LINK_PRIVATE sc-memory)
+target_include_directories(set-agents PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
+```
+
+---
+
+### **4. Define a class in C++ for this agent and specifies class of actions that this agent performs and its program.**
+
+**sc_agent_calculate_set_power.hpp**
+
+```cpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+class ScAgentCalculateSetPower : public ScActionAgent
+{
+public:
+  ScAddr GetActionClass() const override;
+
+  ScResult DoProgram(ScActionEvent const & event, ScAction & action) override;
+};
+```
+
+An agent's class to be implemented must comply with the following requirements:
+
+* It must inherit one of the common classes for implementing agents: 
+  
+  - `template <class TScEvent> class ScAgent`,
+  - or `class ScActionAgent`.
+
+The base class `ScAgent` contains API to implement agents that react to any sc-events. The base class `ScActionAgent` inherits base class `ScAgent` and provides API to implement agents that react to sc-events of initiating sc-action.
+
+* It must override at least methods `ScAddr GetAction() const` and `ScResult DoProgram(ScActionEvent const & event, ScAction & action)`.
+
+* Override methods must be public. Otherwise, you won't be able to build your code because the sc-machine won't be able to call methods on your agent class.
+
+* You can implement other methods in agent's class.
+
+---
+
+### **5. Implement all declared methods of agent's class.**
+
+**sc_agent_calculate_set_power.cpp**
+
+```cpp
+#include "sc_agent_calculate_set_power.hpp"
+
+ScAddr ScAgentCalculateSetPower::GetActionClass() const
+{
+  return m_memoryCtx.HelperFindBySystemIdtf("action_calculate_set_power");
+}
+// You must specify valid action class. In other case, the agent can’t be 
+// subscribed to sc-event.
+
+ScResult ScAgentCalculateSetPower::DoProgram(ScAction & action)
```

Comment:
Out-of-line definition of 'DoProgram' does not match any declaration in 'ScAgentCalculateSetPower'

Reply:
Fixed